tests/lookup/tests.py

```diff
@@ -603,6 +604,27 @@ def test_in_keeps_value_ordering(self):
         query = Article.objects.filter(slug__in=['a%d' % i for i in range(1, 8)]).values('pk').query
         self.assertIn(' IN (a1, a2, a3, a4, a5, a6, a7) ', str(query))
 
+    def test_in_ignores_none(self):
+        with self.assertNumQueries(1) as ctx:
+            self.assertQuerysetEqual(Article.objects.filter(id__in=[None, self.a1.id]), ['<Article: Article 1>'])
+        sql = ctx.captured_queries[0]['sql']
+        self.assertNotIn('NULL', sql)
+
+    def test_in_ignores_solo_none(self):
+        self.assertQuerysetEqual(Article.objects.filter(id__in=[None]), [])
```

Comment:
We can add assertion for number of queries `self.assertNumQueries(0)` because that's the real impact of this optimization.

Reply:
Added